src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -66,44 +69,85 @@ private static void LogHttpRequest(HttpRequestMessage request, Guid loggingReque
+            if (currentActivity != null)
+            {
+                request.Headers.Add(DiagnosticsHandlerLoggingStrings.RequestIdHeaderName, currentActivity.Id);
+                //we expect baggage to be empty or contain a few items
+                using (IEnumerator<KeyValuePair<string, string>> e = currentActivity.Baggage.GetEnumerator())
                 {
-                    long timestamp = Stopwatch.GetTimestamp();
-
-                    if (t.IsFaulted && s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.ExceptionWriteName))
+                    if (e.MoveNext())
                     {
-                        s_diagnosticListener.Write(
-                            DiagnosticsHandlerLoggingStrings.ExceptionWriteName,
-                            new
-                            {
-                                LoggingRequestId = (Guid) s,
-                                Timestamp = timestamp,
-                                Exception = t.Exception,
-                            }
-                        );
+                        var baggage = new List<string>();
+                        do
+                        {
+                            KeyValuePair<string, string> item = e.Current;
+                            baggage.Add(new NameValueHeaderValue(item.Key, item.Value).ToString());
+                        }
+                        while (e.MoveNext());
+                        request.Headers.Add(DiagnosticsHandlerLoggingStrings.CorrelationContextHeaderName, baggage);
                     }
+                }
+            }
 
-                    if (s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.ResponseWriteName))
+            Task<HttpResponseMessage> responseTask = base.SendAsync(request, cancellationToken);
+            try
+            {
+                await responseTask.ConfigureAwait(false);
+            }
+            catch (TaskCanceledException)
+            {
+                //we'll report task status in Activity.Stop
+                throw;
+            }
+            catch (Exception ex)
+            {
+                if (s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.ExceptionEventName))
+                {
+                    //If request was initialy instrumented, Activity.Current has all necessary context for logging
+                    //Request is passed to provide some context if instrumentation was disabled and to avoid
+                    //extensive Activity.Tags usage to tunnel request properties
+                    s_diagnosticListener.Write(DiagnosticsHandlerLoggingStrings.ExceptionEventName, new { Exception = ex, Request = request });
+                }
+                throw;
+            }
+            finally
+            {
+                //always stop activity if it was started
+                if (activity != null)
+                {
+                    activity.SetEndTime(DateTime.UtcNow);
```

Comment:
@pharring -  I suspect ASP.NET used both UtcNow and Stopwatch because up until about a week ago UtcNow was NOT very precise (16 msec).   It is now upgraded to about 1usec, which should be good enough for HTTP request and response durations.   That should be enough.